--- recipes/tasks/task-shr-minimal.bb.org	2010-06-11 23:51:20.815393998 +0200
+++ recipes/tasks/task-shr-minimal.bb	2010-06-11 23:54:15.036598596 +0200
@@ -60,6 +60,7 @@
 	${PN}-x \
 	${PN}-apps \
 	${PN}-gtk \
+	${PN}-nfc \
 "
 
 
@@ -144,3 +145,13 @@
   matchbox-keyboard-im \
 "
 
+RDEPENDS_${PN}-nfc = "\
+  pcsc-lite \
+  libpcsclite \
+  ccid \
+  opensc \
+  libnfc \
+  ifdnfc \
+  ccid-emulator \
+  pace-gui \
+"
