Index: build/build
===================================================================
--- build.orig/build	2012-04-15 20:42:56.183646915 +0200
+++ build/build	2012-04-15 20:43:06.951700312 +0200
@@ -419,7 +419,7 @@
 	OPENSC_COMPONENTS="pcsc"
 	export PKG_CONFIG="true"
 	export MAN2HTML="true"
-	export ZLIB_LIBS="-L${OPENSC_ROOT}/lib -lzdll"
+	#export ZLIB_LIBS="-L${OPENSC_ROOT}/lib -lzdll"
 	export ICONV_CFLAGS="-I${OPENSC_ROOT}/include"
 	export ICONV_LIBS="-L${OPENSC_ROOT}/lib -liconv"
 fi
